src/ints/xms.cpp

```diff
@@ -102,392 +110,752 @@ struct XMS_MemMove{
-		for (Bitu i = 1;i<XMS_HANDLES;i++) 
-			if(!xms_handles[i].free) XMS_FreeMemory(i);
+	if (!section->Get_bool("xms")) {
+		return;
 	}
 
-};
-static XMS* test;
+	// NTS: Disable XMS emulation if CPU type is less than a 286, because
+	// extended memory did not exist until the CPU had enough address lines
+	// to read past the 1MB mark.
+	//
+	// The other reason we do this is that there is plenty of software that
+	// assumes 286+ instructions if they detect XMS services, including but
+	// not limited to:
+	//
+	//      MSD.EXE Microsoft Diagnostics
+	//      Microsoft Windows 3.0
+	//
+	// Not emulating XMS for 8086/80186 emulation prevents the software
+	// from crashing.
+
+	if (CPU_ArchitectureType < CPU_ARCHTYPE_286) {
+		LOG_WARNING("XMS: CPU 80186 or lower lacks address lines needed for XMS, disabling");
+		return;
+	}
+
+	xms.is_available = true;
+	// TODO: read HMA configuration
+
+	BIOS_ZeroExtendedSize(true);
+	DOS_AddMultiplexHandler(xms_multiplex);
+
+	// Place hookable callback in writable memory area
+	xms.callback = RealMake(static_cast<uint16_t>(DOS_GetMemory(0x1) - 1),
+	                        0x10);
+	callbackhandler.Install(&XMS_Handler,
+	                        CB_HOOKABLE,
+	                        Real2Phys(xms.callback),
+	                        "XMS Handler");
+	// pseudocode for CB_HOOKABLE:
+	//	jump near skip
+	//	nop,nop,nop
+	//	label skip:
+	//	callback XMS_Handler
+	//	retf
+
+	for (uint16_t index = 0; index < NumXmsHandles; ++index) {
+		xms.handles[index] = XMS_Block();
+	}
+	xms.handles[0].is_free = false;
+
+	// Set up UMB chain
+	umb.is_available = section->Get_bool("umb");
+	const bool ems_available = GetEMSType(section) > 0;
+	DOS_BuildUMBChain(section->Get_bool("umb"), ems_available);
+
+	// TODO: If implementing CP/M compatibility, mirror the JMP
+	//       instruction in HMA
+}
+
+XMS::~XMS()
+{
+	// Remove upper memory information
+	dos_infoblock.SetStartOfUMBChain(0xffff);
+	if (umb.is_available) {
+		dos_infoblock.SetUMBChainState(0);
+		umb.is_available = false;
+	}
 
-void XMS_ShutDown(Section* /*sec*/) {
-	delete test;	
+	if (!xms.is_available) {
+		return;
+	}
+
+	// Undo biosclearing
+	BIOS_ZeroExtendedSize(false);
+
+	// Remove Multiplex
+	DOS_DelMultiplexHandler(xms_multiplex);
+
+	// Free used memory while skipping the 0 handle
+	for (uint16_t index = 1; index < NumXmsHandles; ++index) {
+		xms.handles[index].lock_count = 0;
+		if (!xms.handles[index].is_free) {
+			xms_free_memory(index);
+		}
+	}
+
+	xms.is_available = false;
+}
+
+// ***************************************************************************
+// Lifecycle
+// ***************************************************************************
+
+static XMS* instance = nullptr;
+
+static void XMS_ShutDown(Section* /* sec */)
+{
+	delete instance;
+	instance = nullptr;
 }
 
 void XMS_Init(Section* sec)
 {
 	assert(sec);
 
-	test = new XMS(sec);
+	if (!instance) {
+		instance = new XMS(sec);
+	}
```

Comment:
```c++
static std::unique_ptr<XMS> instance = {};

// clear it
instance = {};

// assign a new one
instance = std::make_unique<XMS>(sec);
```

The benefit here is the `instance` will tidily delete itself on any early scope closure, which means less to wade through using the leak sanitizer on exceptions or E_Exit scenarios.

(In terms of bloat concerns, unique_ptr's are so minimal that users ask how it can have no size overhead :sweat_smile:  https://stackoverflow.com/questions/13460395/how-can-stdunique-ptr-have-no-size-overhead)

Reply:
OK, done.